llvm/lib/YkIR/YkIRWriter.cpp

```diff
@@ -6,30 +6,120 @@
 
 #include "llvm/ADT/Triple.h"
 #include "llvm/BinaryFormat/ELF.h"
+#include "llvm/IR/Constant.h"
 #include "llvm/IR/Constants.h"
+#include "llvm/IR/InstrTypes.h"
+#include "llvm/IR/Instructions.h"
 #include "llvm/IR/Module.h"
 #include "llvm/MC/MCContext.h"
 #include "llvm/MC/MCSectionELF.h"
 #include "llvm/MC/MCStreamer.h"
+#include "llvm/Support/ErrorHandling.h"
 
 using namespace llvm;
 using namespace std;
 
 namespace {
 
+class SerialiseInstructionException {
+private:
+  string S;
+
+public:
+  SerialiseInstructionException(string S) : S(S) {}
+  string &what() { return S; }
+};
+
 const char *SectionName = ".yk_ir";
 const uint32_t Magic = 0xedd5f00d;
 const uint32_t Version = 0;
 
 enum OpCode {
   Nop = 0,
+  Load,
+  Store,
+  Alloca,
+  Call,
+  GetElementPtr,
+  Branch,
+  ICmp,
+  BinaryOperator,
+  Ret,
+  UnimplementedInstruction = 255, // Will eventually be deleted.
+};
+
+enum OperandKind {
+  Constant = 0,
+  LocalVariable,
+  String,
+};
+
+enum TypeKind {
+  Integer = 0,
+  UnimplementedType = 255, // Will eventually be deleted.
 };
 
+string valueToString(Value *V) {
+  string S;
+  raw_string_ostream SS(S);
+  V->print(SS);
+  return S;
+}
+
+// Get the idex of an element in its parent container.
+template <class C, class E> size_t getIndex(C *Container, E *FindElement) {
+  bool Found = false;
+  size_t Idx = 0;
+  for (E &AnElement : *Container) {
+    if (&AnElement == FindElement) {
+      Found = true;
+      break;
+    }
+    Idx++;
+  }
+  assert(Found);
+  return Idx;
+}
+
+#define GENERIC_INST_SERIALISE(LLVM_INST, LLVM_INST_TYPE, YKIR_OPCODE)         \
+  if (isa<LLVM_INST_TYPE>(LLVM_INST)) {                                        \
+    serialiseInstGeneric(LLVM_INST, YKIR_OPCODE);                              \
+    return;                                                                    \
+  }
+
 class YkIRWriter {
 private:
   Module &M;
   MCStreamer &OutStreamer;
 
+  vector<llvm::Type *> Types;
+  vector<llvm::Constant *> Constants;
+
+  // Return the index of the LLVM type `Ty`, inserting it a new entry if
+  // necessary.
+  size_t typeIndex(Type *Ty) {
+    vector<Type *>::iterator Found = std::find(Types.begin(), Types.end(), Ty);
+    if (Found != Types.end()) {
+      return std::distance(Types.begin(), Found);
+    }
+    size_t Idx = Types.size();
+    Types.push_back(Ty);
+    return Idx;
+  }
+
```

Comment:
~~it~~

Reply:
Fixed in dd5424b160ceb1f8f820b9c238821377624bac19